### PR TITLE
RETURN DISTINCT execution plan operation (#359)

### PR DESCRIPTION
**Pull request by rafie**
- Introduced xxhash library
- Added Record_Hash64 function
- Implemented OpDistinct

### CODE REVIEW

File: src/execution_plan/ops/op_distinct.c [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    self->op.free = DistinctFree;
+
+    return (OpBase*)self;
+}
+
+Record DistinctConsume(OpBase *opBase) {
+    OpDistinct *self = (OpDistinct*)opBase;
+    OpBase *child = self->op.children[0];
+
+    while(true) {
```
Comment (swilly22) [inline src/execution_plan/ops/op_distinct.c]: Wouldn't you prefer:
while((r = child->consume(child))) {
    unsigned long long const hash = Record_Hash64(r); ....
}

File: src/execution_plan/ops/op_distinct.c [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        Record r = child->consume(child);
+        if(!r) return NULL;
+
+        unsigned long long const hash = Record_Hash64(r);
+        int is_new = TrieMap_Add(self->trie, (char *) &hash, sizeof(hash), NULL, TrieMap_DONT_CARE_REPLACE);
+        if(is_new)
```
Comment (swilly22) [inline src/execution_plan/ops/op_distinct.c]: One liner
if(is_new) return r;

File: src/execution_plan/record.c [hunk 3]
```diff
@@ -111,6 +113,90 @@ size_t Record_ToString(const Record r, char **buf, size_t *buf_cap) {
...
+
+    res = XXH64_reset(&state, 0);
+    assert(res != XXH_ERROR);
+    
+    for(int i = 0; i < rec_len; ++i) {
+        Entry e = r[i];
+        switch(e.type) {
+        case REC_TYPE_NODE:
+            entity.type = GraphEntityType_NODE;
+            entity.id = ENTITY_GET_ID(Record_GetGraphEntity(r, i));
```
Comment (swilly22) [inline src/execution_plan/record.c]: Replace Record_GetGraphEntity  with Record_GetNode

File: src/execution_plan/record.c [hunk 4]
```diff
@@ -111,6 +113,90 @@ size_t Record_ToString(const Record r, char **buf, size_t *buf_cap) {
...
+            data = &entity;
+            len = sizeof(entity);
+            break;
+            
+        case REC_TYPE_EDGE:
+            entity.type = GraphEntityType_EDGE;
+            entity.id = ENTITY_GET_ID(Record_GetGraphEntity(r, i));
```
Comment (swilly22) [inline src/execution_plan/record.c]: Replace Record_GetGraphEntity  with Record_GetEdge

File: src/execution_plan/record.c [hunk 5]
```diff
@@ -111,6 +113,90 @@ size_t Record_ToString(const Record r, char **buf, size_t *buf_cap) {
...
+            switch (si.type) {
+            case T_NULL:
+                data = &_null;
+                len = sizeof(_null);
+                break;
+                
+            case T_STRING:
+            case T_CONSTSTRING:
+                data = si.stringval;
+                len = si.stringval ? strlen(si.stringval) : 0;
```
Comment (swilly22) [inline src/execution_plan/record.c]: Why can't we just do:
`len = strlen(si.stringval)`

File: tests/flow/test_distinct.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
+from redis_base import RedisGraphTestBase
+
+class ReturnDistinctFlowTest_1(RedisGraphTestBase):
+    @classmethod
+    def createCommands(cls):
```
Comment (swilly22) [inline tests/flow/test_distinct.py]: why createCommand in ReturnDistinctFlowTest_1 differs from createCommand in ReturnDistinctFlowTest_2 ?

File: tests/flow/test_distinct.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+        # limit+distinct
+        q = self.query("MATCH (p:PARENT)-[:HAS]->(:CHILD) RETURN DISTINCT p LIMIT 2")
+        self.assertEqual(q, [['Stevie'], ['Mike']])
+        
+        # order+limit+distinct
+        q = self.query("MATCH (p:PARENT)-[:HAS]->(:CHILD) RETURN DISTINCT p ORDER BY p.name LIMIT 2")
+        self.assertEqual(q, [['James'], ['Mike']])
+
+class ReturnDistinctFlowTest_2(RedisGraphTestBase):
```
Comment (swilly22) [inline tests/flow/test_distinct.py]: Do we really need ReturnDistinctFlowTest_2?

File: src/execution_plan/record.c (hunk 3 above)
Comment (rafie) [reply src/execution_plan/record.c]: This reverts a previous change request: https://github.com/RedisLabsModules/RedisGraph/pull/379#discussion_r260709192

File: src/execution_plan/record.c (hunk 4 above)
Comment (rafie) [reply src/execution_plan/record.c]: This reverts a previous change request: https://github.com/RedisLabsModules/RedisGraph/pull/379#discussion_r260709964

File: src/execution_plan/record.c (hunk 5 above)
Comment (rafie) [reply src/execution_plan/record.c]: I prefer not to assume si.stringval != NULL.

File: tests/flow/test_distinct.py (hunk 6 above)
Comment (rafie) [reply tests/flow/test_distinct.py]: The only difference is the order of the resultset. These tests take note of that difference. However, if the order will change in the future and the test fails, what should we do? My intuition is to keep the 2nd test, even though the change in order is not directly related to the DISTINCT operation.

File: tests/flow/test_distinct.py (hunk 7 above)
Comment (rafie) [reply tests/flow/test_distinct.py]: We don't really need it. However, it documents the difference in ordering of the resultset coming from a different create query. It not directly related to the DISTINCT operation.

File: src/execution_plan/ops/op_distinct.c (hunk 1 above)
Comment (rafie) [reply src/execution_plan/ops/op_distinct.c]: Well, I'm not sure whether the following variant is actually clearer than the original one.

Record r;
while((r = child->consume(child))) {
    unsigned long long const hash = Record_Hash64(r);
    int is_new = TrieMap_Add(self->trie, (char *) &hash, sizeof(hash), NULL, TrieMap_DONT_CARE_REPLACE);
    if(is_new) return r;
    Record_Free(r);
}
return NULL;

File: src/execution_plan/record.c (hunk 5 above)
Comment (swilly22) [reply src/execution_plan/record.c]: if an SIValue is of type T_STRING or T_CONSTSTRING be assure that si.stringval isn't NULL.

**Pull request merged**